### PR TITLE
node-lifecycle-controller: improve processPod test-coverage

### PR DESCRIPTION
**Pull request by carlory**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

`processPod` is not covered by tests. 

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

**nodelifecycle**  The coverage has now been increased to cover 69.3% of the statements vs the current  64.9% of statements coverage

```
(base) ➜  kubernetes git:(KEP-3902-test) make test WHAT=./pkg/controller/nodelifecycle KUBE_COVER=y
...
+++ [1009 18:35:59] Saving coverage output in '/tmp/k8s_coverage/20241009-183559'
+++ [1009 18:35:59] Running tests with code coverage and with -race
ok  	k8s.io/kubernetes/pkg/controller/nodelifecycle	20.639s	coverage: 64.9% of statements

DONE 46 tests in 81.961s
+++ [1009 18:37:23] Combined coverage report: /tmp/k8s_coverage/20241009-183559/combined-coverage.html
(base) ➜  kubernetes git:(master) git checkout -
Switched to branch 'KEP-3902-test'
Your branch is up to date with 'origin/KEP-3902-test'.
(base) ➜  kubernetes git:(KEP-3902-test) make test WHAT=./pkg/controller/nodelifecycle KUBE_COVER=y
...
+++ [1009 18:38:57] Saving coverage output in '/tmp/k8s_coverage/20241009-183857'
+++ [1009 18:38:57] Running tests with code coverage and with -race
ok  	k8s.io/kubernetes/pkg/controller/nodelifecycle	12.254s	coverage: 69.3% of statements

DONE 47 tests in 19.573s
+++ [1009 18:39:17] Combined coverage report: /tmp/k8s_coverage/20241009-183857/combined-coverage.html
```

```diff
(base) ➜  kubernetes git:(KEP-3902-test)  diff <(go tool cover -func=/tmp/k8s_coverage/20241009-183559/combined-coverage.out) <(go tool cover -func=/tmp/k8s_coverage/20241009-183857/combined-coverage.out)
7c7
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:307:	NewNodeLifecycleController		42.0%
---
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:307:	NewNodeLifecycleController		47.8%
17c17
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1087:	podUpdated				0.0%
---
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1087:	podUpdated				80.0%
19c19
< k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1115:	processPod				0.0%
---
> k8s.io/kubernetes/pkg/controller/nodelifecycle/node_lifecycle_controller.go:1115:	processPod				57.7%
29c29
< total:											(statements)				64.9%
---
> total:											(statements)				69.3%
```

**nodelifecycle/scheduler** The coverage has now been increased to cover 92.6% of the statements vs the current  85.3%  of statements coverage

```
(base) ➜  kubernetes git:(master) make test WHAT=./pkg/controller/nodelifecycle/scheduler KUBE_COVER=y
Makefile:195: warning: undefined variable `''
+++ [1009 18:45:56] Set GOMAXPROCS automatically to 8
+++ [1009 18:45:56] Normalizing Go targets
WARNING: ulimit -n (files) should be at least 1000, is 256, may cause test failure
+++ [1009 18:45:56] Saving coverage output in '/tmp/k8s_coverage/20241009-184556'
+++ [1009 18:45:56] Running tests with code coverage and with -race
ok  	k8s.io/kubernetes/pkg/controller/nodelifecycle/scheduler	1.358s	coverage: 85.3% of statements

DONE 8 tests in 2.056s
+++ [1009 18:45:58] Combined coverage report: /tmp/k8s_coverage/20241009-184556/combined-coverage.html
(base) ➜  kubernetes git:(master) git checkout -
Switched to branch 'KEP-3902-test'
Your branch is up to date with 'origin/KEP-3902-test'.
(base) ➜  kubernetes git:(KEP-3902-test) make test WHAT=./pkg/controller/nodelifecycle/scheduler KUBE_COVER=y
Makefile:195: warning: undefined variable `''
+++ [1009 18:46:07] Set GOMAXPROCS automatically to 8
+++ [1009 18:46:07] Normalizing Go targets
WARNING: ulimit -n (files) should be at least 1000, is 256, may cause test failure
+++ [1009 18:46:07] Saving coverage output in '/tmp/k8s_coverage/20241009-184607'
+++ [1009 18:46:07] Running tests with code coverage and with -race
ok  	k8s.io/kubernetes/pkg/controller/nodelifecycle/scheduler	1.269s	coverage: 92.6% of statements

DONE 9 tests in 1.615s
+++ [1009 18:46:08] Combined coverage report: /tmp/k8s_coverage/20241009-184607/combined-coverage.html
```

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go [hunk 1]
```diff
@@ -3557,3 +3568,138 @@ func Test_isNodeExcludedFromDisruptionChecks(t *testing.T) {
...
+				Clientset: fake.NewSimpleClientset(&v1.PodList{Items: []v1.Pod{*testutil.NewPod("pod0", "node0")}}),
+			},
+			pod:                     testutil.NewPod("pod0", "node0"),
+			monitorNodeHealth:       true,
+			expectedPodStatusUpdate: true,
+		},
+	}
+
+	_, ctx := ktesting.NewTestContext(t)
+	for i, item := range table {
```
Comment (bart0sh) [inline pkg/controller/nodelifecycle/node_lifecycle_controller_test.go:3665]: Can you use `t.Run` as in the `TestMonitorNodeHealth ` and `TestMonitorNodeHealthMarkPodsNotReadyRetry` tests?

Comment (carlory) [reply pkg/controller/nodelifecycle/node_lifecycle_controller_test.go]: updated.

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go [hunk 2]
```diff
@@ -3557,3 +3568,141 @@ func Test_isNodeExcludedFromDisruptionChecks(t *testing.T) {
...
+				},
+				Clientset: fake.NewSimpleClientset(&v1.PodList{Items: []v1.Pod{*testutil.NewPod("pod0", "node0")}}),
+			},
+			pod:                     testutil.NewPod("pod0", "node0"),
+			monitorNodeHealth:       true,
+			expectedPodStatusUpdate: true,
+		},
+	}
+
+	_, ctx := ktesting.NewTestContext(t)
```
Comment (bart0sh) [inline pkg/controller/nodelifecycle/node_lifecycle_controller_test.go:3665]:
```suggestion
	ctx := ktesting.Init(t)
```

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go [hunk 3]
```diff
@@ -3557,3 +3568,141 @@ func Test_isNodeExcludedFromDisruptionChecks(t *testing.T) {
...
+				},
+				Clientset: fake.NewSimpleClientset(&v1.PodList{Items: []v1.Pod{*testutil.NewPod("pod0", "node0")}}),
+			},
+			pod:                     testutil.NewPod("pod0", "node0"),
+			monitorNodeHealth:       true,
+			expectedPodStatusUpdate: true,
+		},
+	}
+
+	ctx := ktesting.Init(t)
```
Comment (bart0sh) [inline pkg/controller/nodelifecycle/node_lifecycle_controller_test.go:3665]: This API belongs to `k8s.io/kubernetes/test/utils/ktesting`

Comment (carlory) [reply pkg/controller/nodelifecycle/node_lifecycle_controller_test.go]: @bart0sh updated and renamed ctx to tCtx like other controllers' unit tests.

**Comment by carlory** [comment]
@bart0sh updated.

File: pkg/controller/nodelifecycle/node_lifecycle_controller_test.go [hunk 4]
```diff
@@ -785,13 +796,13 @@ func TestMonitorNodeHealth(t *testing.T) {
 
 	for testName, tt := range tests {
 		t.Run(testName, func(t *testing.T) {
-			_, ctx := ktesting.NewTestContext(t)
+			tCtx := ktesting.Init(t)
```
Comment (SergeyKanzhelev) [inline pkg/controller/nodelifecycle/node_lifecycle_controller_test.go:799]: @pohly which context is the right one to use? I am lost between available contexts for tests.

Comment (SergeyKanzhelev) [inline pkg/controller/nodelifecycle/node_lifecycle_controller_test.go:799]: should we initialize the context outside the loop? Like you do in newly added `TestProcessPodMarkPodNotReady`?

Comment (pohly) [reply pkg/controller/nodelifecycle/node_lifecycle_controller_test.go:799]: Both are functionally equivalent as far as contextual logging goes. tCtx offers some additional APIs, but that's a bit experimental.

Comment (pohly) [reply pkg/controller/nodelifecycle/node_lifecycle_controller_test.go:799]: The closer a context is to the actual test code the better, i.e. innermost. That way the output is associated with `testName` in this case, not with the parent test.

Comment (SergeyKanzhelev) [reply pkg/controller/nodelifecycle/node_lifecycle_controller_test.go:799]: Thanks. @carlory please change the new method `TestProcessPodMarkPodNotReady` then to initialize context inside the loop

Comment (carlory) [reply pkg/controller/nodelifecycle/node_lifecycle_controller_test.go:799]: updated.

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm
/approve

**Pull request merged**